Spigot-Server-Patches/0387-Add-Velocity-IP-Forwarding-Support.patch

```diff
@@ -0,0 +1,294 @@
+diff --git a/src/main/java/net/minecraft/server/LoginListener.java b/src/main/java/net/minecraft/server/LoginListener.java
+index 5778a5201..d7284e3b5 100644
+--- a/src/main/java/net/minecraft/server/LoginListener.java
++++ b/src/main/java/net/minecraft/server/LoginListener.java
+@@ -43,6 +43,10 @@ public class LoginListener implements PacketLoginInListener, ITickable {
+     private SecretKey loginKey;
+     private EntityPlayer l;
+     public String hostname = ""; // CraftBukkit - add field
++    // Paper start - Velocity support
++    private static final MinecraftKey VELOCITY_CHANNEL = new MinecraftKey("velocity", "player_info");
++    private int velocityLoginMessageId = -1;
++    // Paper end
+ 
+     public LoginListener(MinecraftServer minecraftserver, NetworkManager networkmanager) {
+         this.g = LoginListener.EnumProtocolState.HELLO;
+@@ -187,6 +191,14 @@ public class LoginListener implements PacketLoginInListener, ITickable {
+             this.g = LoginListener.EnumProtocolState.KEY;
+             this.networkManager.sendPacket(new PacketLoginOutEncryptionBegin("", this.server.E().getPublic(), this.e));
+         } else {
++            // Paper start - Velocity support
++            if (com.destroystokyo.paper.PaperConfig.velocitySupport) {
++                this.velocityLoginMessageId = java.util.concurrent.ThreadLocalRandom.current().nextInt();
++                PacketLoginOutCustomPayload packet = new PacketLoginOutCustomPayload(this.velocityLoginMessageId, VELOCITY_CHANNEL, new PacketDataSerializer(io.netty.buffer.Unpooled.EMPTY_BUFFER));
++                this.networkManager.sendPacket(packet);
++                return;
++            }
++            // Paper end
+             // Spigot start
+             // Paper start - Cache authenticator threads
+             authenticatorPool.execute(new Runnable() {
+@@ -278,6 +290,12 @@ public class LoginListener implements PacketLoginInListener, ITickable {
+     public class LoginHandler {
+ 
+         public void fireEvents() throws Exception {
++                            // Paper start - Velocity support
++                            if (LoginListener.this.velocityLoginMessageId == -1 && com.destroystokyo.paper.PaperConfig.velocitySupport) {
++                                disconnect("This server requires you to connect with Velocity.");
++                                return;
++                            }
++                            // Paper end
+                             String playerName = i.getName();
+                             java.net.InetAddress address = ((java.net.InetSocketAddress) networkManager.getSocketAddress()).getAddress();
+                             java.util.UUID uniqueId = i.getId();
+@@ -325,6 +343,55 @@ public class LoginListener implements PacketLoginInListener, ITickable {
+     // Spigot end
+ 
+     public void a(PacketLoginInCustomPayload packetloginincustompayload) {
++        // Paper start - Velocity support
```

Comment:
I think all the logic that can be extracted should be extracted to its own class: `io.papermc.paper.proxy.VelocityProxy` or so.

Reply:
The only code I think we can extract is the logic that begins the login process in the thread pool executor, which is repeated three times in the time in this PR (and two before).